df_to_azure/export.py

```diff
@@ -291,25 +301,69 @@ def set_upload_name(self, folder: str, method: str) -> str:
         ----------
         folder: str
             the folder name, derived from the df_to_azure parameter 'schema'
-        method: str
-            upload_method.
 
         Returns
         -------
         name: str
             the folder + filename structure for uploading the dataset.
         """
-        if method == "create":
+        if self.method in ("create", "upsert"):
             name = f"{folder}/{self.tablename}.parquet"
-        elif method == "append":
+        elif self.method == "append":
             name = f"{folder}/{self.tablename}/{self.tablename}_{datetime.now().strftime('%Y%m%d%H%M%S')}.parquet"
         else:
-            raise ValueError(f"No valid method given: {method}. choose create or append.")
+            allow_list = ["create", "append", "upsert"]
+            raise ValueError(f"No valid method given: {self.method}. choose from {', '.join(allow_list)}.")
         return name
 
-    def run(self):
+    def upsert(self, df_existing: pd.DataFrame):
+        """
+        Perform insert or update on a pandas DataFrame.
 
-        text_stream = self.df.to_parquet()
+        The new values of self.df will be updated in df_existing,
+        and new rows in self.df will be appended to df_existing.
+
+        1. We read the existing parquet file from storage
+        2. We do the upsert with the given dataframe
+        3. We upload and overwrite the parquet file on storage with the updated dataframe
+
+        Parameters
+        ----------
+        df_existing: pd.DataFrame
+            The dataframe which is already on Azure Storage and has to be updated.
+
+        Returns
+        -------
+        result: pd.DataFrame
+            Updated dataframe to be uploaded.
+        """
+        # check if columns are equal
+
+        diff_cols = self.df.columns.symmetric_difference(df_existing.columns)
+        if diff_cols.any():
+            err_msg = (
+                f"When performing upsert, column names must be equal. " f"Difference in columns: {', '.join(diff_cols)}"
+            )
+            raise ValueError(err_msg)
+
+        # set indices with id columns
+        df_existing = df_existing.set_index(self.id_field)
+        self.df = self.df.set_index(self.id_field)
+        # perform upsert
+        self.df = self.df.combine_first(df_existing)
```

Comment:
Als ik naar de documentatie kijk van `combine_first()` dan update deze alleen null values, maar deze update deze geen velden waar al een waarde in staat? En volgens mij is dat wel wat je belooft bij upsert?

Reply:
En je tests falen niet omdat deze verkeerd om staan, nu update je self.df met df_existing, maar je moet df_existing updaten met self.df. Dus volgens mij zit de functionaliteit er nu nog niet goed in.

Reply:
Kan je een reproduceerbaar fout maken? Want als ik de huidige logica volg, dan gaat het goed. ID's met nieuwe waardes worden geupdate en de huidige regels worden ongemoeid gelaten.

![image](https://user-images.githubusercontent.com/34067903/149331451-6b848737-dbb4-454d-ace4-c4dcdbf78494.png)

Reply:
![image](https://user-images.githubusercontent.com/81814590/149336697-767e7893-b1db-4eff-baae-e6303fe835cd.png)
De laatste code is wat je moet doen (bestaand updaten met de nieuwe), alleen het resultaat daarvan is niet juist. De huidige manier update de nieuwe met de oude, maar er update niks terwijl de uitkomst logischerwijs dan nog wel de nieuwe df is.